factorio_noir/render.py

```diff
@@ -71,15 +69,40 @@ def matrix(self, saturation, brightness):
 DEFAULT_COLORSPACE = ColorSpace(X=0.3086, Y=0.6094, Z=0.0820)
 
 
-def apply_transforms(image, saturation, brightness):
+def apply_transforms(image, treatment):
     """Apply the needed transformations to the given image."""
     img_alpha = image.getchannel("A")
     img_rgb = image.convert("RGB")
 
-    color_space = DEFAULT_COLORSPACE.matrix(saturation, brightness)
+    color_space = DEFAULT_COLORSPACE.matrix(treatment.saturation, treatment.brightness)
 
     img_converted = img_rgb.convert("RGB", color_space)
+
+    for x_index, x_tile in enumerate(treatment.tiling):
+        y_max = len(x_tile)
+        for y_index, tile_strength in enumerate(x_tile):
+            if tile_strength == 1:
+                continue
+
+            # TODO: Double check for off by one errors here
+            bounding_box = (
+                int(image.width * x_index / len(treatment.tiling)),
+                int(image.height * y_index / len(x_tile)),
+                int(image.width * (x_index + 1) / len(treatment.tiling)),
+                int(image.height * (y_index + 1) / len(x_tile)),
+            )
+
+            image_box = img_rgb.crop(bounding_box)
+            # converted_box = img_converted.crop(bounding_box)
+
+            # blended_box = image_box.blend(converted_box, tile_strength)
+
+            # TODO: why doesn't blend work?
+            assert tile_strength == 0
+            blended_box = image_box
+
+            img_converted.paste(blended_box, bounding_box)
+
     img_converted.putalpha(img_alpha)
```

Comment:
Would tiling work after the alpha is added? If so maybe having a dedicated function would make the code more readable

Reply:
Both the colour matrix & the tiling works on an RGB image. After everything is done, the unmodified alpha layer is re-added back in.

I might try playing around with this to see if I can make it a little nicer